ntoskrnl/inbv/inbv.c

```diff
@@ -762,6 +769,147 @@ NtDisplayString(IN PUNICODE_STRING DisplayString)
+        Index = 32;
+    }
+    X = ProgressBarLeft + 2;
+    Y = ProgressBarTop + 2;
+    InbvReleaseLock();
+
+    while (InbvDisplayState == INBV_DISPLAY_STATE_OWNED && PltRotBarStatus != 3)
+    {
+        InbvAcquireLock();
+
+        if (RotBarSelection == RB_SQUARE_CELLS)
+        {
+            if (Index >= 3)
+            {
+                /* Fill previous bar position */
+                VidSolidColorFill(X + ((Index - 3) * 8), Y, (X + ((Index - 3) * 8)) + 8 - 1, Y + 9 - 1, 0);
+            }
+            if (Index < Total - 1)
+            {
+                /* Draw the progress bar bit */
+                if (Index < 2)
+                {
+                    /* Appearing from the left */
+                    VidBufferToScreenBlt(RotBarBuffer + 8 * (2 - Index) / 2, X, Y, 22 - 8 * (2 - Index), 9, 24);
+                }
+                else if (Index >= Total - 3)
+                {
+                    /* Hiding to the right */
+                    VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22 - 8 * (4 - (Total - Index)), 9, 24);
+                }
+                else
+                {
+                    VidBufferToScreenBlt(RotBarBuffer, X + ((Index - 2) * 8), Y, 22, 9, 24);
+                }
+            }
+            Index = (Index + 1) % Total;
+        }
+        else if (RotBarSelection == RB_PROGRESS_BAR)
+        {
+            /* Right part */
+            VidBufferToScreenBlt(RotLineBuffer, Index, 474, 640 - Index, 6, 640);
+            if (Index > 0)
+            {
+                /* Left part */
+                VidBufferToScreenBlt(RotLineBuffer + (640 - Index) / 2, 0, 474, Index - 2, 6, 640);
+            }
+            Index = (Index + 32) % 640;
+        }
+
+        InbvReleaseLock();
+
+        /* Wait for a bit */
+        KeDelayExecutionThread(KernelMode, FALSE, &Delay);
+    }
+    PsTerminateSystemThread(STATUS_SUCCESS);
+}
+
+VOID
+NTAPI
+INIT_FUNCTION
+InbvRotBarInit(VOID)
+{
+    NTSTATUS Status;
+    HANDLE ThreadHandle = NULL;
+
+    if (!InbvBootDriverInstalled || RotBarSelection == RB_UNSPECIFIED)
+        return;
+
+    RotBarThread = NULL;
+    PltRotBarStatus = 0;
+
+    Status = PsCreateSystemThread(&ThreadHandle,
+                                  SYNCHRONIZE,
+                                  NULL,
+                                  NULL,
+                                  NULL,
+                                  InbvRotationThread,
+                                  NULL);
+
+    if (NT_SUCCESS(Status))
+    {
+        Status = ObReferenceObjectByHandle(ThreadHandle,
+                                           SYNCHRONIZE,
+                                           PsThreadType,
+                                           KernelMode,
+                                           &RotBarThread,
+                                           NULL);
+        if (NT_SUCCESS(Status))
+        {
+            KeSetPriorityThread(RotBarThread, HIGH_PRIORITY);
```

Comment:
Why does this need high priority? It's a graphical gimmick that shouldn't slow down the actual boot process if at all possible.

Reply:
Some kernel drivers (`SCSIPORT`, for example, as figured out by @HBelusca) use `KeStallExecutionProcessor` that make animation thread totally hang. This was added to improve it a bit, but I don't know whether its really neccessary.